File: migrate/go/target.go
package migrate

import (
	"fmt"
	"strconv"
	"strings"
)

// ResolveTargets resolves a tern-style destination spec into the ordered list of absolute
// target versions to migrate to (design.md §6/§9). current is the version presently
// recorded; n is the highest available sequence. The spec grammar:
//
//	"last" or ""  → migrate to N (the default; the dominant startup case)
//	"<integer>"   → migrate to that absolute version
//	"+N"          → migrate up N steps   (current + N)
//	"-N"          → migrate down N steps (current - N)
//	"-+N"         → redo the last N: migrate down N, then back up N ([current-N, current])
//
// Every resolved target is range-checked against 0 … N; an out-of-range result is an error.
// Most specs resolve to a single target; the redo form resolves to two, applied in order.
// Relative-grammar resolution is the caller's concern (design.md §9 — typically a CLI); the
// library's MigrateTo takes only an absolute target.
func ResolveTargets(spec string, current, n int) ([]int, error) {
	spec = strings.TrimSpace(spec)
	if spec == "" || spec == "last" {
		return []int{n}, nil
	}

	// Redo: "-+N" (down N, then up N).
	if rest, ok := strings.CutPrefix(spec, "-+"); ok {
		steps, err := strconv.Atoi(rest)
		if err != nil || steps < 0 {
			return nil, fmt.Errorf("bad destination %q: expected -+N", spec)
		}
		down := current - steps
		if err := checkRange(down, n); err != nil {
			return nil, err
		}
		return []int{down, current}, nil
	}

	// Relative up/down: "+N" / "-N".
	if strings.HasPrefix(spec, "+") || strings.HasPrefix(spec, "-") {
		delta, err := strconv.Atoi(spec) // Atoi handles the leading sign
		if err != nil {
			return nil, fmt.Errorf("bad destination %q: expected +N, -N, -+N, an integer, or last", spec)
		}
		target := current + delta
		if err := checkRange(target, n); err != nil {
			return nil, err
		}
		return []int{target}, nil
	}

	// Absolute integer.
	target, err := strconv.Atoi(spec)
	if err != nil {
		return nil, fmt.Errorf("bad destination %q: expected +N, -N, -+N, an integer, or last", spec)
	}
	if err := checkRange(target, n); err != nil {
		return nil, err
	}
	return []int{target}, nil
}

func checkRange(target, n int) error {
	if target < 0 || target > n {
		return &BadVersion{Version: target, N: n, Whence: "target"}
	}
	return nil
}

File: migrate/rust/src/target.rs
//! Resolving tern-style destination specs to absolute target versions (design.md §6/§9).

use crate::error::MigrateError;

/// Resolve a tern-style destination `spec` into the ordered list of absolute target versions
/// to migrate to (design.md §6/§9). `current` is the version presently recorded; `n` is the
/// highest available sequence. The grammar:
///
/// | spec | meaning |
/// |---|---|
/// | `"last"` or `""` | migrate to `N` (the default) |
/// | `"<integer>"` | migrate to that absolute version |
/// | `"+N"` | migrate up `N` steps (`current + N`) |
/// | `"-N"` | migrate down `N` steps (`current - N`) |
/// | `"-+N"` | redo the last `N`: down `N`, then back up `N` (`[current-N, current]`) |
///
/// Every resolved target is range-checked against `0 … N`. Relative-grammar resolution is the
/// caller's concern (design.md §9 — typically a CLI); the library's
/// [`migrate_to`](crate::Migrator::migrate_to) takes only an absolute target.
pub fn resolve_targets(spec: &str, current: u32, n: u32) -> Result<Vec<u32>, MigrateError> {
    let spec = spec.trim();
    if spec.is_empty() || spec == "last" {
        return Ok(vec![n]);
    }

    // Redo: "-+N" (down N, then back up N). Checked before the "-N" case.
    if let Some(rest) = spec.strip_prefix("-+") {
        let steps: u32 = rest.parse().map_err(|_| bad_dest(spec))?;
        let down = (current as i64) - (steps as i64);
        check_range(down, n)?;
        return Ok(vec![down as u32, current]);
    }

    // Relative up/down: "+N" / "-N".
    if let Some(rest) = spec.strip_prefix('+') {
        let steps: u32 = rest.parse().map_err(|_| bad_dest(spec))?;
        let target = (current as i64) + (steps as i64);
        check_range(target, n)?;
        return Ok(vec![target as u32]);
    }
    if let Some(rest) = spec.strip_prefix('-') {
        let steps: u32 = rest.parse().map_err(|_| bad_dest(spec))?;
        let target = (current as i64) - (steps as i64);
        check_range(target, n)?;
        return Ok(vec![target as u32]);
    }

    // Absolute integer.
    let target: i64 = spec.parse().map_err(|_| bad_dest(spec))?;
    check_range(target, n)?;
    Ok(vec![target as u32])
}

fn check_range(target: i64, n: u32) -> Result<(), MigrateError> {
    if target < 0 || target > n as i64 {
        return Err(MigrateError::BadVersion {
            version: target,
            n,
            whence: "target",
        });
    }
    Ok(())
}

fn bad_dest(spec: &str) -> MigrateError {
    MigrateError::Load(format!(
        "bad destination {spec:?}: expected +N, -N, -+N, an integer, or last"
    ))
}
